test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs

```diff
@@ -0,0 +1,159 @@
+            this.runExpirationTime = runExpirationTime;
+            this.cancellationToken = Preconditions.CheckNotNull(cancellationToken, nameof(cancellationToken));
+            this.edgeHubRestartedTime = edgeHubRestartedTime;
+            this.restartSequenceNumber = restartSequenceNumber;
+            this.logger = Preconditions.CheckNotNull(logger, nameof(logger));
+        }
+
+        public async Task StartAsync()
+        {
+            (DateTime dmCompletedTime, HttpStatusCode dmStatusCode) = await this.SendDirectMethodAsync(
+                Settings.Current.DeviceId,
+                Settings.Current.DirectMethodTargetModuleId,
+                this.dmModuleClient,
+                Settings.Current.DirectMethodName,
+                this.runExpirationTime,
+                this.cancellationToken,
+                this.logger).ConfigureAwait(false);
+
+            TestResultBase dmTestResult = new EdgeHubRestartDirectMethodResult(
+                Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString(),
+                DateTime.UtcNow,
+                Settings.Current.TrackingId,
+                this.batchId,
+                Interlocked.Read(ref this.directMethodCount).ToString(),
+                this.edgeHubRestartedTime,
+                dmCompletedTime,
+                dmStatusCode,
+                this.restartSequenceNumber);
+
+            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
+            await ModuleUtil.ReportTestResultUntilSuccessAsync(
+                reportClient,
+                this.logger,
+                dmTestResult,
+                this.cancellationToken).ConfigureAwait(false);
+        }
+
+        async Task<Tuple<DateTime, HttpStatusCode>> SendDirectMethodAsync(
+            string deviceId,
+            string targetModuleId,
+            ModuleClient moduleClient,
+            string directMethodName,
+            DateTime runExpirationTime,
+            CancellationToken cancellationToken,
+            ILogger logger)
+        {
+            while ((!cancellationToken.IsCancellationRequested) && (DateTime.UtcNow < runExpirationTime))
+            {
+                try
+                {
+                    // Direct Method sequence number is always increasing regardless of sending result.
+                    Interlocked.Increment(ref this.directMethodCount);
+                    MethodRequest request = new MethodRequest(
+                        directMethodName,
+                        Encoding.UTF8.GetBytes($"{{ \"Message\": \"Hello\", \"DirectMethodCount\": \"{Interlocked.Read(ref this.directMethodCount).ToString()}\" }}"),
+                        Settings.Current.SdkOperationTimeout,
+                        Settings.Current.SdkOperationTimeout);
+                    MethodResponse result = await moduleClient.InvokeMethodAsync(deviceId, targetModuleId, request);
```

Comment:
So we are actually waiting for a client to connect and respond? What if that client takes longer to connect/respond? Will that cause the test to fail?

Reply:
We intentionally set the SDK timeout to be 20ms, if that timeout happens, it catch an exception and retry